driftbase/api/players/counters.py

```diff
@@ -152,88 +156,62 @@ def _patch(self, player_id):
             if missing_keys:
                 result[name] = "Missing keys: %s" % ",".join(missing_keys)
                 continue
 
-            counter_type = entry.get("counter_type", DEFAULT_COUNTER_TYPE).lower()
+            counter_type = update.get("counter_type", DEFAULT_COUNTER_TYPE).lower()
             if counter_type not in LEGAL_COUNTER_TYPES:
                 result[name] = "Illegal counter type. Expecting: %s" % ",".join(LEGAL_COUNTER_TYPES)
                 continue
 
-            counter_type = entry.get("counter_type", DEFAULT_COUNTER_TYPE)
-            counters[name] = {"counter_type": counter_type}
-
-            context_id = int(entry.get("context_id", 0))
-
-            redis_start_time = time.time()
-            # write the counter into redis as well. I will refactor the db calls out
-            # and use only redis here in the future
-            value = int(entry["value"])
-            key = 'counters:%s:%s:%s:%s:%s' % (name, counter_type, player_id, timestamp, context_id)
-            # expire in 1 day
-            ex = 86400
-            if counter_type == "absolute":
-                g.redis.set(key, value, expire=ex)
+            is_absolute = counter_type == COUNTER_TYPE_ABSOLUTE
+            value = float(update["value"])
+            context_id = int(update.get("context_id", 0))
+            # ensure that multiple updates all get applied
+            # theoretically these should be individual entries, if the client flushes at a low rate,
+            # but since we don't trust the client's time stamp, it's better to merge them for now
+            update_entry = counter_updates.get(name)
+            if update_entry:
+                if is_absolute:
+                    update_entry["value"] = value
+                else:
+                    update_entry["value"] += value
             else:
-                g.redis.incr(key, value, expire=ex)
-            log.info("Added %s to redis in %.2fsec", name, time.time() - redis_start_time)
-
-        for counter_name, counter_info in counters.items():
-            counter_id = get_or_create_counter_id(counter_name, counter_info["counter_type"])
-            player_counter = get_or_create_player_counter(counter_id, player_id)
-            counter_info["player_counter"] = player_counter
-            counter_info["counter_id"] = counter_id
+                counter_updates[name] = dict(counter_type=counter_type, value=value,
+                                             context_id=context_id,
+                                             is_absolute=is_absolute,
+                                             timestamp=timestamp
+                                             )
 
-        g.db.commit()
+        counter_ids = []
+        counters = batch_get_or_create_counters([(k, v["counter_type"]) for k, v in counter_updates.items()])
+        for (counter_id, name) in counters:
+            counter_updates[name].update(dict(counter_id=counter_id))
```

Comment:
`counter_updates[name]["counter_id"] = counter_id` is clearer and faster.

Reply:
Right, I was updating multiple properties inititally